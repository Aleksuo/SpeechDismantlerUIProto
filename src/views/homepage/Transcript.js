import React, { Component } from 'react'
import { Paper } from '@material-ui/core'
import PropTypes from 'prop-types'
import { CSSTransitionGroup } from 'react-transition-group'


import Sentence from './Sentence'

/**
 * A component that displays the transcript elements
 * @author Aleksi Suoranta
 */
class Transcript extends Component {
    constructor(props) {
        super(props)
        this.state = {
            currentPlayback: 0,
            isPlaying: false
        }
        this.audio = React.createRef()
        //scroll to these references
        this.transcriptEnd = React.createRef()
        this.currentlyPlaying = React.createRef()
    }

    /**
     * Binds event handlers to the html audio element here
     * @this {Transcript}
     * @author Aleksi Suoranta
     */
    componentDidMount() {
        this.audio.current.ontimeupdate = () => {
            this.setState({ currentPlayback: this.audio.current.currentTime })
        }

        this.audio.current.onloadedmetadata = () => {
            //A hack for a chrome bug: https://stackoverflow.com/questions/38443084/how-can-i-add-predefined-length-to-audio-recorded-from-mediarecorder-in-chrome/39971175
            this.audio.current.currentTime = 10000000 * Math.random()
        }

        this.audio.current.onplaying = () => {
            this.setState({ isPlaying: true })
        }

        this.audio.current.onpaused = () => {
            this.setState({ isPlaying: false })
        }
    }

    /**
     * Scroll the transcript on updates when needed
     * @this {Transcript}
     * @author Aleksi Suoranta
     */
    componentDidUpdate = () => {
        if (!this.audio.current.paused) {
            this.scrollToCurrentPlayback()
        } else {
            this.scrollToBottom()
        }
    }

    /**
     * Scrolls to the end of the transcript marked by reference div
     * @this {Transcript}
     * @author Aleksi Suoranta
     */
    scrollToBottom = () => {
        this.transcriptEnd.current.scrollIntoView()
    }

    /**
     * Scrolls to the sentence that is currently played, marked by reference div.
     * @this {Transcript}
     * @author Aleksi Suoranta
     */
    scrollToCurrentPlayback = () => {
        if(this.currentlyPlaying.current){
            this.currentlyPlaying.current.scrollIntoView({ behavior: 'smooth' })
        }
    }

    /**
     * Event handler for clicking sentences, starts playback on click.
     * @this {Transcript}
     * @author Aleksi Suoranta
     * @param {number} time
     */
    onSentenceClick = (time) => {
        this.audio.current.currentTime = (time) / 1000
        this.audio.current.play()
    }

    render = () => {
        const { transcript, blobUrl, wordColor } = this.props
        const cur = this.state.currentPlayback * 1000
        const items = transcript.map((sentence, idx, arr) => {
            const next = arr[idx+1] ? arr[idx+1].startTime : Number.MAX_VALUE
            return (
                <Sentence
                    wordColor={wordColor}
                    key={idx}
                    isCurrent={(sentence.startTime <= cur) && (next >= cur) && this.state.isPlaying}
                    sentence={sentence}
                    onClick={() => this.onSentenceClick(sentence.startTime)}
                    ref={this.currentlyPlaying}

                />)
        })

        return (

            <div>
                <Paper elevation={1} style={{ maxHeight: "30vh", height: "30vh", overflow: "auto" }}>
                    <CSSTransitionGroup
                        transitionName="example"
                        transitionEnterTimeout={500}
                        transitionLeaveTimeout={300}>
                        <div>
                            <span>{items}</span>
                            <div ref={this.transcriptEnd} />
                        </div>

                    </CSSTransitionGroup>
                </Paper>


                <div style={{ marginTop: "7px" }}>
                    <audio ref={this.audio} src={blobUrl} controls type="audio/ogg" style={{ width: "100%" }}></audio>
                </div>
            </div>
        )

    }

}


Transcript.propTypes = {
    transcript: PropTypes.array,
    blobUrl: PropTypes.string
}

export default Transcript