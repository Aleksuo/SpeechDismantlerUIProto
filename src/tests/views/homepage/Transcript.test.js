/* eslint-disable */
import React from "react"
import Enzyme, {shallow, mount,render} from "enzyme"
import Transcript from "../../../views/HomePage/Transcript"
import Adapter from "enzyme-adapter-react-16"

Enzyme.configure({adapter: new Adapter()})

describe("Transcript component", () =>{
    test("renders", () =>{ 
        // for some reason using mount causes a memory leak here and refs are not supported in shallow
        const wrapper = render(<Transcript transcript={[]}/>) 
        expect(wrapper).toMatchSnapshot()
    })
})