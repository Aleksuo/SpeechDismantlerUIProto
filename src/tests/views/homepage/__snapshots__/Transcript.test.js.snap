// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transcript component renders 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiPaper-root-1 MuiPaper-elevation1-4 MuiPaper-rounded-2",
          "style": "max-height:30vh;height:30vh;overflow:auto",
        },
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "attribs": Object {},
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "span",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                  Object {
                    "attribs": Object {},
                    "children": Array [],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "span",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "style": "margin-top:7px",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "controls": "",
                "style": "width:100%",
                "type": "audio/ogg",
              },
              "children": Array [],
              "name": "audio",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "controls": undefined,
                "style": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "controls": undefined,
                "style": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
      Object {
        "attribs": Object {
          "style": "margin-top:7px",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "controls": "",
              "style": "width:100%",
              "type": "audio/ogg",
            },
            "children": Array [],
            "name": "audio",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "controls": undefined,
              "style": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "controls": undefined,
              "style": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "MuiPaper-root-1 MuiPaper-elevation1-4 MuiPaper-rounded-2",
            "style": "max-height:30vh;height:30vh;overflow:auto",
          },
          "children": Array [
            Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "attribs": Object {},
                  "children": Array [
                    Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "span",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": Object {
                        "attribs": Object {},
                        "children": Array [],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": [Circular],
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                    Object {
                      "attribs": Object {},
                      "children": Array [],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": Object {
                        "attribs": Object {},
                        "children": Array [],
                        "name": "span",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": [Circular],
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {},
                  "x-attribsPrefix": Object {},
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
